### PR TITLE
feat(gateway)!: implement foundational logic for gas fees

### LINKED ISSUE #2: Add CI
Bare-minimum:
- Typescript linter
- FunC compiler
- Jest tests runner
- Some DevSecOps tools

### PR DESCRIPTION
**Pull request by swift1337**
# Changes

- [x] Make gas fees deterministic for all `op`
- [x] Estimate gas usage, add stats export to CSV
- [x] Add `int calculate_gas_fee(int op) method_id` getter for tx cost estimation 
- [x] Improve TSS signature payload.
- [x] Update readme
- [x] Come up w/ a strategy for non-existent withdrawal addresses
- [x] Setup CI
- [x] Additional test cases
- [x] Add CLI wrapper with `make tx`
- [x] Restrict withdrawals to Gateway itself

Closes #2, #6, #9, #16

### CODE REVIEW

File: contracts/common/state.fc [hunk 1]
```diff
@@ -10,9 +10,6 @@ global int state::deposits_enabled;
 ;; Total locked TON in the gateway
 global int state::total_locked;
 
-;; Total fees collected by the gateway
-global int state::fees;
```
Comment (swift1337) [inline contracts/common/state.fc:14]: It was removed due to redundancy.

File: contracts/common/gas.fc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+;; https://docs.ton.org/develop/howto/blockchain-configs#param-20-and-21
+cell get_gas_config_param(int wc) inline {
+    if (wc == -1) {
+        return config_param(20);
+    } else {
+        return config_param(21);
+    }
+}
+
+(slice, (int, int)) load_gas_flat_pfx(slice param) inline {
```
Comment (lumtis) [inline contracts/common/gas.fc:14]: Would be great to add a bit more documentation on these functions

File: contracts/gateway.fc [hunk 3]
```diff
@@ -25,9 +26,13 @@ const op::authority::update_tss = 202;
...
 
-;; GAS FEE =========================================
-;; Let's use const for now and refine this later
-const deposit_gas_fee = 10000000; ;; 0.01 TON
+;; GAS FEES =========================================
+;; These constants are empirically determined based on tests
+;; and slighly bumped to cover all other non-gas fees (e.g. storage, fwd_fee, etc.)
+const gas::deposit = 10000;
+const gas::deposit_and_call = 13000;
+const gas::authority = 20000;
```
Comment (lumtis) [inline contracts/gateway.fc:34]: What is authority? Admin functions?

File: contracts/gateway.fc [hunk 4]
```diff
@@ -25,9 +26,13 @@ const op::authority::update_tss = 202;
 const op::authority::update_code = 203;
 const op::authority::update_authority = 204;
 
-;; GAS FEE =========================================
-;; Let's use const for now and refine this later
-const deposit_gas_fee = 10000000; ;; 0.01 TON
+;; GAS FEES =========================================
+;; These constants are empirically determined based on tests
+;; and slighly bumped to cover all other non-gas fees (e.g. storage, fwd_fee, etc.)
```
Comment (lumtis) [inline contracts/gateway.fc:31]: What slighly bumped means in this context? I think we should set a comfortable buffer to ensure we don't have overpayment issue

File: contracts/gateway.fc [hunk 5]
```diff
@@ -344,3 +355,31 @@ int seqno() method_id {
 
     return state::seqno;
 }
+
+;; returns tx fee for the specified operation
+int calculate_gas_fee(int op) method_id {
```
Comment (lumtis) [inline contracts/gateway.fc:360]: It seems like this function is not used in the code?

Comment (swift1337) [reply contracts/gateway.fc:360]: it's a getter function that might be used via RPC to estimate tx fee in TON

File: contracts/gateway.fc (hunk 4 above)
Comment (swift1337) [reply contracts/gateway.fc:31]: tx fee = gas_fee + additional fees (storage fee, forward fee, etc...). Based on the test cases I made tx fee be approximated to `gas_price(gas::operation_abc)`. 

This allows each operation to have a cost "ceiling" i.e. `real total fees` <= `gas_price(gas::operation_abc)`

**Review by lumtis — APPROVED** [review]

**Pull request merged**